### DISCUSSION TITLE
AutoFixture.AutoNSubstitute and customizations?

### DISCUSSION THREAD

**Discussion by Ergamon** [AutoFixture/AutoFixture#1231] [Q&A]
Propably another different APIs working together problem, but perhaps you have some ideas.

Is there any way to make these 2 systems work together?

Let´s keep it simple and leave out all the xUnit stuff.

```
public interface IPerson
{
  int Id { get; }
  string Name { get; }
}
```

If we look at this simple interface I can use AutoFixture in this way to generate an IPerson instance:

```
var fixture = new Fixture();
var customization = new AutoNSubstituteCustomization()
{
  ConfigureMembers = true
};
fixture.Customize(customization);

var person = fixture.Create<IPerson>();
```
By doing this I get a substitute for IPerson with some random values for the 2 properties.

So if e.g. I dont want a random number for the Id, how can I customize this?

What I tried so far:
```
fixture.Customize<IPerson>(c => c.Without(p => p.Id));
```
Sadly this fails basically twice. First AutoFixture tells me that it is not possible to set a non writeable property.

Even if I change the interface to have a setter on the Id property, it then fails with the message that AutoFixture cannot create the interface. Which seems logical, cause I basically overwrote the AutoNSubstituteCustomization ...

This factory problem can be resolved by using FromFactory, something like this:
```
fixture.Customize<IPerson>(
  c => c.FromFactory(() => Substitute.For<IPerson>())
  .With(p => p.Id, 42));
```
Basically doing the substitution for myself, but still having the readonly property problem.

I can of course now change the factory method with code like this:
```
fixture.Customize<IPerson>(c => c.FromFactory(() =>
{
  var p = Substitute.For<IPerson>();
  p.Name.Returns(nameof(p.Name) + fixture.Create<string>());
  return p;
}));
```

But that counterfeights the whole idea of AutoFixture. Cause now I have to eventually change this every time the interface changes.
Have I overseen anything or is this just a limitation?
